# Methods to fit an overparameterized linear model

Basic algebra teaches us that a system of equations having N equations with N unknown parameters can be solved-- with a unique solution. But perfect solvability isn't the only situation. Define p as the number of unknown parameters.  Suppose there are N equations. Then there are two other cases:
* underparameterized: When p < N, then there are no solutions (or the system can be re-written as fewer equations).
* overparameterized: When p > N, there are many possible solutions.

Here I focus on the overparameterized case.  How do we choose a single solution?

# Setting up the example

But first, consider the simplest case of N equations with N unknowns, which is one equation with one unknown.  Here is an example:

`10 = 2 x`

That has one solution, namely `x = 5`.  (Note that we solve this by taking the product of 10 by the inverse of 2, which is 1/2.)

An underparameterize case would mean more equations than unknowns.  So let's add one more equation:

```
10 = 2 x
10 = 5 x
```

Clearly, no solution is perfect and we will need a way to select the "least bad" solution.  That way is usually linear regression with ordinary least squares (OLS).

Now consider the overparameterized case, which means there are more unknowns than equations, i.e. p > N. In the simplest case, there is one equation with TWO unknowns.  Let's add one more parameter to the above equation

`10 = 2 x + 4 y`

Clearly, there are many possible solutions and we will need a way to select one.  I will show a way by minimizing the L2 norm, equivalent to calculating the Moore-Penrose inverse.


The important point here is that the interpolation case is the only one with no freedom or wiggle room.  There is ONE solution and you are stuck with it, regardless of how well it generalizes.

# Libraries

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_rows', 14)

# Example underparameterized non-solutions


Recall our under-parameterized system of equations:

```
10 = 2 x
10 = 5 x
```

Let's consider some possible values of x and how far the equations would make them from the actual goal of 10.

In [2]:
x = np.arange(1.0, 6.0, 0.5)
df = pd.DataFrame.from_dict({'x':x, '2x':2*x, '2x-10':(2*x-10), '5x':5*x, '5x-10':(5*x-10), 'RSS':((2*x-10)**2+(5*x-10)**2)})
print(df.to_string(index=False))

  x   2x  2x-10   5x  5x-10    RSS
1.0  2.0   -8.0  5.0   -5.0  89.00
1.5  3.0   -7.0  7.5   -2.5  55.25
2.0  4.0   -6.0 10.0    0.0  36.00
2.5  5.0   -5.0 12.5    2.5  31.25
3.0  6.0   -4.0 15.0    5.0  41.00
3.5  7.0   -3.0 17.5    7.5  65.25
4.0  8.0   -2.0 20.0   10.0 104.00
4.5  9.0   -1.0 22.5   12.5 157.25
5.0 10.0    0.0 25.0   15.0 225.00
5.5 11.0    1.0 27.5   17.5 307.25


So if we chose the lowest-RSS value for x, it would be x=2.5, with an RSS of 31.25.

OLS stands for "ordinary least squares" and this method of picking an under-parameterized solution can sometimes generalize well.  It equivalent to root mean square but without the root-- giving the same ordering though.

I don't assess generalization error here.  This was just a train-only error, not including a test set.

But just for kicks, to start thinking about generalization, let's add the L2 norm to our table.

In [3]:
x = np.arange(1.0, 6.0, 0.5)
df = pd.DataFrame.from_dict({'x':x, 'RSS':((2*x-10)**2+(5*x-10)**2), 'L2':(x**2), 'RSS+2*L2':((2*x-10)**2+(5*x-10)**2) + 2*x**2})
print(df.to_string(index=False))

  x    RSS    L2  RSS+2*L2
1.0  89.00  1.00     91.00
1.5  55.25  2.25     59.75
2.0  36.00  4.00     44.00
2.5  31.25  6.25     43.75
3.0  41.00  9.00     59.00
3.5  65.25 12.25     89.75
4.0 104.00 16.00    136.00
4.5 157.25 20.25    197.75
5.0 225.00 25.00    275.00
5.5 307.25 30.25    367.75


# Example overparameterized solutions

We consider some possible values of x and determine what y would match those.  Note that because all of these solutions lead to a value of 10, there is no error and  RSS is always zero, so we cannot use the RSS to select one.

In [4]:
x = np.arange(0.5, 1.3, 0.1)
df = pd.DataFrame.from_dict({'x':x, 'y':(10-2*x)/4})
df['RSS'] = [0] * len(x)
print(df.to_string(index=False))

  x    y  RSS
0.5 2.25    0
0.6 2.20    0
0.7 2.15    0
0.8 2.10    0
0.9 2.05    0
1.0 2.00    0
1.1 1.95    0
1.2 1.90    0


## Choosing a solution with the min L2 norm

One possible way to choose a solution is to choose the one that minimizes the l2 norm of the parameters, meaning `x^2 + y^2`.  It is possible to solve this analystically, using the Moore-Penrose inverse, the matrix equivalent of taking the inverse of a number (1/2 vs. 2).

But here let's skip the math and just calculate the L2 values for our possible solutions and choose the minimum.

In [5]:
df["L2"] = df["x"]**2 + df["y"]**2
print(df.to_string(index=False))

  x    y  RSS     L2
0.5 2.25    0 5.3125
0.6 2.20    0 5.2000
0.7 2.15    0 5.1125
0.8 2.10    0 5.0500
0.9 2.05    0 5.0125
1.0 2.00    0 5.0000
1.1 1.95    0 5.0125
1.2 1.90    0 5.0500


The minimum L2, namely of 5.00, is attained at:
`(x=1.0, y=2.0)`

So let's choose that solution.

Just for reference, let me also add the residual sum of squares (RSS) to the table.

In [6]:
df = pd.DataFrame.from_dict({'x':x, 'y':(10-2*x)/4})
df["RSS"] = [0] * len(df)
df["L2"] = df["x"]**2 + df["y"]**2
print(df.to_string(index=False))

  x    y  RSS     L2
0.5 2.25    0 5.3125
0.6 2.20    0 5.2000
0.7 2.15    0 5.1125
0.8 2.10    0 5.0500
0.9 2.05    0 5.0125
1.0 2.00    0 5.0000
1.1 1.95    0 5.0125
1.2 1.90    0 5.0500


# Regularization

Now suppose we were using ridge regression.  Regularization does not require an exact solution so we are no longer restricted to `2x + 5y = 10`, that is, no longer require RSS to be zero.  And we minimize a weighted sum of RSS and the L2 norm:

`RSS + 𝛼 L2`

So for 𝛼=1.0 we would get the chart below.

In [7]:
x = np.arange(0.0, 4.0, 0.5)
df = pd.DataFrame.from_dict({'x':x, 'y':(10-2*x)/4})
df["predict"] = [10] * len(df)
df["RSS"] = [0] * len(df)
df["L2"] = df["x"]**2 + df["y"]**2
df

x     y  predict  RSS       L2
0  0.0  2.50       10    0   6.2500
1  0.5  2.25       10    0   5.3125
2  1.0  2.00       10    0   5.0000
3  1.5  1.75       10    0   5.3125
4  2.0  1.50       10    0   6.2500
5  2.5  1.25       10    0   7.8125
6  3.0  1.00       10    0  10.0000
7  3.5  0.75       10    0  12.8125

In [8]:
alpha = 1.0

In [9]:
more = pd.DataFrame.from_dict({"x":[0.95], "y":[1.90], "predict":[2*0.95+4*1.9], "RSS":[(10-(2+0.95+4*1.9))**2], 'L2':[(0.95**2+1.9**2)]})
df2 = pd.concat([df, more])
df2.sort_values(by='x', inplace=True)
df2["ridge"] = df2["RSS"] +  alpha*df2["L2"]

In [10]:
print(df2.to_string(index=False))

   x    y  predict    RSS      L2   ridge
0.00 2.50     10.0 0.0000  6.2500  6.2500
0.50 2.25     10.0 0.0000  5.3125  5.3125
0.95 1.90      9.5 0.3025  4.5125  4.8150
1.00 2.00     10.0 0.0000  5.0000  5.0000
1.50 1.75     10.0 0.0000  5.3125  5.3125
2.00 1.50     10.0 0.0000  6.2500  6.2500
2.50 1.25     10.0 0.0000  7.8125  7.8125
3.00 1.00     10.0 0.0000 10.0000 10.0000
3.50 0.75     10.0 0.0000 12.8125 12.8125


We see that with the ridge metric, which is RSS + L2, the minimum is achieved for (x=0.95, y=1.90).

# The former interpolating regime

Interestingly, when we have the interpolating regime of 1 equation with 1 unknown, Ridge regression gives some freedom in choosing (x, y) values because an exact solution with an RSS of 0 is no longer the minimum.

Recall we were solving

`10 = 2 x`

If we require an exact solution (RSS=0), then x=5.0 is the only possible value.  Let's see what happens when we try some other values.

In [11]:
x = np.arange(3.0, 7.0, 0.5)
df = pd.DataFrame.from_dict({'x':x, 'predict':2*x, 'RSS':((2*x-10)**2), 'L2':x**2})
df["ridge"] = df["RSS"] + alpha*df["L2"]
print(df.to_string(index=False))

  x  predict  RSS    L2  ridge
3.0      6.0 16.0  9.00  25.00
3.5      7.0  9.0 12.25  21.25
4.0      8.0  4.0 16.00  20.00
4.5      9.0  1.0 20.25  21.25
5.0     10.0  0.0 25.00  25.00
5.5     11.0  1.0 30.25  31.25
6.0     12.0  4.0 36.00  40.00
6.5     13.0  9.0 42.25  51.25


A ridge regularizer would favor x=4.0 over x=5.0.  Although x=4.0 incurs some error (RSS is not zero), its L2 norm is smaller, meaning it is less likely to be fitting noise.